File: reader_iface.go
type Reader interface {
	AccountReader
	Balance(addr solana.PublicKey) (uint64, error)
	SlotHeight() (uint64, error)
	LatestBlockhash() (*rpc.GetLatestBlockhashResult, error)
	ChainID() (string, error)
	GetFeeForMessage(msg string) (uint64, error)
	GetLatestBlock() (*rpc.GetBlockResult, error)
}

File: reader_iface.rs
use solana_client::rpc;
use std::error::Error;

pub trait Reader: AccountReader {
    fn balance(&self, addr: solana_sdk::pubkey::Pubkey) -> Result<u64, Box<dyn Error>>;
    fn slot_height(&self) -> Result<u64, Box<dyn Error>>;
    fn latest_blockhash(&self) -> Result<rpc::GetLatestBlockhashResult, Box<dyn Error>>;
    fn chain_id(&self) -> Result<String, Box<dyn Error>>;
    fn get_fee_for_message(&self, msg: &str) -> Result<u64, Box<dyn Error>>;
    fn get_latest_block(&self) -> Result<rpc::GetBlockResult, Box<dyn Error>>;
}

